controllers/crds/cninode_controller.go

```diff
@@ -0,0 +1,194 @@
+// permissions and limitations under the License.
+
+package crds
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/amazon-vpc-resource-controller-k8s/apis/vpcresources/v1alpha1"
+	ec2API "github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/aws/ec2/api/cleanup"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/config"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/k8s"
+	"github.com/aws/amazon-vpc-resource-controller-k8s/pkg/utils"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// CNINodeReconciler reconciles a CNINode object
+type CNINodeReconciler struct {
+	client.Client
+	Scheme      *runtime.Scheme
+	Context     context.Context
+	Log         logr.Logger
+	EC2Wrapper  ec2API.EC2Wrapper
+	K8sAPI      k8s.K8sWrapper
+	ClusterName string
+	VPCID       string
+}
+
+//+kubebuilder:rbac:groups=vpcresources.k8s.aws,resources=cninodes,verbs=get;list;watch;create;update;patch;
+
+// Reconcile handles CNINode create/update/delete events
+// Reconciler will add the finalizer and cluster name tag if it does not exist and finalize on CNINode on deletion to clean up leaked resource on node
+func (r *CNINodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cniNode := &v1alpha1.CNINode{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("CNINode is deleted", "CNINode", req.NamespacedName)
+		}
+		// Ignore not found error; CNINode will be re-created on node reconcile if the node is not terminated
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Add cluster name tag if it does not exist
+	clusterNameTagKey := fmt.Sprintf(config.ClusterNameTagKeyFormat, r.ClusterName)
+	val, ok := cniNode.Spec.Tags[clusterNameTagKey]
+	if !ok || val != config.ClusterNameTagValue {
+		cniNodeCopy := cniNode.DeepCopy()
+		if len(cniNodeCopy.Spec.Tags) != 0 {
+			cniNodeCopy.Spec.Tags[clusterNameTagKey] = config.ClusterNameTagValue
+		} else {
+			cniNodeCopy.Spec.Tags = map[string]string{
+				clusterNameTagKey: config.ClusterNameTagValue,
+			}
+		}
+
+		fmt.Println("patching CNINode to add tags")
+		if err := r.Client.Patch(ctx, cniNodeCopy, client.MergeFrom(cniNode)); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to patch CNINode %v", err)
+		}
+
+		// Re-fetch the CNINode after updating the Tag so we have the latest state of the CRD before updating it again in later operations
+		if err := r.Client.Get(ctx, req.NamespacedName, cniNode); err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
I am not sure if I follow what this if check wants to do here? You `Get` the object and `Patch` it. Why do you need `Get` it again here?

Reply:
Since currently it is updating both Tag and Finalizer field on the same reconcile, we need to re-fetch the latest version of the object else it fails with `object has been modified, please apply your changes to the latest version and try again`. 

Do you mean we just return after patching the tag so we add the finalizer on next reconcile triggered for CNINode? That should work too, but it takes two reconciles to get to desired state